src/common/framework/logging/test_case_recorder.ts

```diff
@@ -48,35 +63,63 @@ export class TestCaseRecorder {
     if (!this.debugging) {
       return;
     }
-    this.logs.push(new LogMessageWithStack('DEBUG', ex, false));
+    const logMessage = new LogMessageWithStack('DEBUG', ex);
+    logMessage.printStack = false;
+    this.logImpl(LogSeverity.Pass, logMessage);
+  }
+
+  skipped(ex: SkipTestCase): void {
+    this.logImpl(LogSeverity.Skip, new LogMessageWithStack('SKIP', ex));
   }
 
   warn(ex: Error): void {
-    this.setState(PassState.warn);
-    this.logs.push(new LogMessageWithStack('WARN', ex));
+    this.logImpl(LogSeverity.Warn, new LogMessageWithStack('WARN', ex));
   }
 
-  fail(ex: Error): void {
-    this.setState(PassState.fail);
-    this.logs.push(new LogMessageWithStack('FAIL', ex));
+  expectationFailed(ex: Error): void {
+    this.logImpl(LogSeverity.ExpectFailed, new LogMessageWithStack('EXPECTATION FAILED', ex));
   }
 
-  skipped(ex: SkipTestCase): void {
-    this.setState(PassState.skip);
-    this.logs.push(new LogMessageWithStack('SKIP', ex));
+  validationFailed(ex: Error): void {
+    this.logImpl(LogSeverity.ValidationFailed, new LogMessageWithStack('VALIDATION FAILED', ex));
   }
 
   threw(ex: Error): void {
     if (ex instanceof SkipTestCase) {
       this.skipped(ex);
       return;
     }
-
-    this.setState(PassState.fail);
-    this.logs.push(new LogMessageWithStack('EXCEPTION', ex));
+    this.logImpl(LogSeverity.ThrewException, new LogMessageWithStack('EXCEPTION', ex));
   }
 
-  private setState(state: PassState): void {
-    this.state = Math.max(this.state, state);
+  private logImpl(level: LogSeverity, logMessage: LogMessageWithStack): void {
+    // Deduplicate errors with the exact same stack
+    if (logMessage.stack) {
+      const seen = this.seenStacks.get(logMessage.stack);
```

Comment:
"seen" makes it seem like `seenStacks` is a map of stack to boolean. Perhaps rename it to "seenLogMessage"